WalletWasabi/WabiSabi/Crypto/CredentialIssuer.cs

```diff
@@ -200,14 +212,17 @@ public CredentialsResponse HandleRequest(CredentialsRequest registrationRequest)
 			var macs = credentials.Select(x => x.Mac);
 			var response = new CredentialsResponse(macs, proofs);
 
-			// Register the serial numbers to prevent credential reuse.
-			foreach (var presentation in presented)
+			return delegate
```

Comment:
something like this?

```
class PreparedCredentialsResponse {
    private CredentialIssuer Issuer { get; }
    private CredentialsRequest Request { get; }
    private CredentialsResponse Response { get; }
    
    ...
    
    public CredentialsResponse Commit()
    {
        ... // delegate body

        return Response;
    }
}
```

Reply:
I didn't see this conversation before posting my comment. This is what I would like to avoid because it adds a new type, it is more OOP and doesn't improve anything, quite the opposite because now it is easier to a mistake. For example, I should not be able to create an instance of this class in any way and that's not possible (or really hard to achieve without adding more trash). 

In my opinion if we change the code it has to be to make it simpler and not more fragile and verbose.

Reply:
Ii generally feel the same which is why i preferred a delegate, but I don't feel strongly about it. I didn't see this response before pushing 1ffdf3a07cfe30dd9eef6d7de6bda211bac14bb1 but at any rate I leave it to you two to decide, i'm happy with keeping 1ffdf3a07cfe30dd9eef6d7de6bda211bac14bb1 or with discarding it, and i'm open to different approaches as well but agree that the sealed internal class approach is adding too much boilerplate/friction and making everything public is just an invitation for headaches.